File: KMP算法/KMP.java
import java.util.*;

public class KMP {
    public static int[] getNext(String mainString) {
        int length = mainString.length();
        int[] next = new int[length];

        for (int ind = 1; ind < length; ++ind) {
            for (int j = ind; j > 0;) {
                j = next[j - 1];
                if (mainString.charAt(j) == mainString.charAt(ind)) {
                    next[ind] = j + 1;
                    break;
                }
            }
        }
        return next;
    }

    public static int find(String mainString, String patternString) {
        int[] next = getNext(patternString);

        int i = 0, j = 0;
        while (i <= mainString.length() - patternString.length()) {
            if (mainString.charAt(i) != patternString.charAt(j)) {
                // backtrace
                if (j == 0)
                    i++;
                else
                    j = next[j - 1];
                continue;
            }
            i++;
            j++;
            if (j == patternString.length())
                return i - j;
        }
        return -1;
    }

    public static void main(String[] args) {
        System.out.println(find("what the fuck", "e "));
        System.out.println(find("what the fuck", "the1"));
    }
}

File: KMP算法/kmp.py
# coding: utf8


def get_next(pattern):
    next_array = [0] * len(pattern)
    for ind in range(1, len(pattern)):
        j = ind
        while j > 0:
            j = next_array[j - 1]
            if pattern[ind] == pattern[j]:
                next_array[ind] = j + 1
                break
    return next_array


def kmp(main_string, pattern):
    # 计算 next 数组
    next_array = get_next(pattern)

    i = j = 0
    while i < len(main_string):
        if main_string[i] == pattern[j]:
            if j == len(pattern) - 1:
                return i - j
            i = i + 1
            j = j + 1
            continue
        if j == 0:
            i = i + 1
        else:
            j = next_array[j - 1]


if __name__ == "__main__":
    print(kmp("bbaaababaabbaabb", "abaabb"))
